lib/file-list.js

```diff
@@ -131,16 +131,16 @@ class FileList {
 
   // Check if we are currently refreshing
   _isRefreshing () {
-    return this._refreshing.isPending()
+    return this._refreshing && this._refreshing.isPending()
   }
 
   // Do the actual work of refreshing
   _refresh () {
     const buckets = this.buckets
     const matchedFiles = new Set()
 
-    let promise
-    promise = Promise.map(this._patterns, (patternObject) => {
+    let lastCompletedRefresh = this._refreshing
```

Comment:
Please read the comment below ;-)

Reply:
Do you think it's possible to write unit tests that fails when you replace `let lastCompletedRefresh = this._refreshing` with `let lastCompletedRefresh` - like it was before ?

Reply:
Well maybe I'm confused.  To me
```js
let lastCompletedRefresh = Promise.map(this._patterns, (patternObject) => {...})
   .then(() => { 
        if (this._refreshing !== lastCompletedRefresh) {
          return this._refreshing
        }
     })
...
```
means that once this statement executes, lastCompletedRefresh is set to the promise returned by the RHS.   In the promise callback the value is unambiguous.  However this line will cause the test to fail with `ReferenceError: lastCompletedRefresh is not defined`. 
If we write
```js
let lastCompletedRefresh = null
lastCompletedRefresh = Promise.map(this._patterns, (patternObject) => {...})
   .then(() => { 
        if (lastCompletedRefresh === null) {
          throw new Error('how can this be')
        }
     })
...
```
Then one of the unit tests throws "how can this be".  I don't understand how the `then` call back can run before the promise assignment.

So, no, but I also don't understand the unit tests behavior now.

(This long chain of promises at different levels is to hard to read. IMO it should be refactored to make the chains clearer).

Reply:
> Then one of the unit tests throws "how can this be". I don't understand how the then call back can run before the promise assignment.

Because of something like [this](https://github.com/karma-runner/karma/blob/master/test/unit/file-list.spec.js#L565), which causes promise callbacks to be executed synchronously (and therefore before the assignment).

I'm working on simplifying this piece now.